community/rfcs/techniques/proposals/20200205-weight-clustering-proposal.md

```diff
@@ -0,0 +1,79 @@
+
+| Status        | Pending                                       |
+:---------------|:----------------------------------------------|
+| **Author(s)** | Mohamed Nour Abouelseoud (mohamednour.abouelseoud@arm.com), Aron Virginas-Tar (aron.virginas-tar@arm.com), Anton Kachatkou (anton.kachatkou@arm.com)|
+| **Sponsor**   | Alan Chiao (alanchiao@google.com)             |
+| **Updated**   | 2020-03-16                                    |
+
+We propose to implement the technique of weight clustering on top of the Keras API.
+
+## Motivation and Overview
+
+Clustering, or weight sharing, brings improvements via model compression. The technique revolves around reducing the number of unique weights in a model, thus facilitating storage and over-the-wire size reduction. Further framework and hardware support can also unlock memory footprint improvements.
+
+Clustering can be combined with pruning and quantization, reducing the network’s storage requirements even further &mdash; see the paper linked [below](#proposed-design) for more details.
+
+## Experimental Results
+
+The table below summarizes the compression rate and top-1 accuracy achieved when clustering the weights of different models using linear centroid initialization.
+
+NOTE: Further networks, e.g. MobileNetV1, are being investigated. The table will be extended to include these once the results become available.
+
+<table>
+  <tr>
+    <th rowspan=2>Model</th>
+    <th rowspan=2>Dataset</th>
+    <th colspan=2>Original</th>
+    <th colspan=3>Clustered</th>
+  </tr>
+  <tr>
+    <th>Top-1 accuracy</th>
+    <th>Size of compressed .tflite</th>
+    <th># of clusters</th>
+    <th>Top-1 accuracy</th>
+    <th>Size of compressed .tflite</th>
+  </tr>
+  <tr>
+    <td>ConvNet</td>
+    <td>MNIST</td>
+    <td>99.40%</td>
+    <td>0.6 MB</td>
+    <td>32</td>
+    <td>98.78%</td>
+    <td>0.1 MB</td>
+  </tr>
+  <tr>
+    <td>MobileNetV2</td>
+    <td>ImageNet</td>
+    <td>72.29%</td>
+    <td>13.0 MB</td>
+    <td>32</td>
+    <td>69.33%</td>
+    <td>2.6 MB</td>
+  </tr>
+  <tr>
+    <td>DS-CNN-L</td>
+    <td>Speech Commands v0.02</td>
+    <td>95.03%</td>
+    <td>1.5 MB</td>
+    <td>32</td>
+    <td>94.71%</td>
+    <td>0.3 MB</td>
+  </tr>
+</table>
+
+NOTE: *Size of compressed .tflite* refers to the size of the zipped .tflite file obtained from the model through the following process:
+1. Serialize the Keras model into .h5 file
+2. Convert the .h5 file into .tflite using `TFLiteConverter.from_keras_model_file()`
+3. Compress the .tflite file into a zip
+
+Zipping the .tflite file is relevant in this context in order to highlight the compression benefits. Clustering does not reduce the size of the weights per se, but it renders them suitable for compression via Huffman coding and similar techniques. In consequence, though the clustered .tflite file will have similar size to the non-clustered one, their zipped versions will greatly differ in size &mdash; as shown by the figures above.
+
```

Comment:
For others, please also include the note of what you've noticed so far with clustering not working well in conjunction with batchnorm folding and per-channel quantization. 

We might be seeing it in Ruomei's end-to-end example, where we go from 8x to 11x compression after applying post-training quantization, as opposed to something like 8x  to 32x.

Reply:
Indeed, these known issues should be mentioned somewhere in the document. I will add a note ASAP.

Reply:
@alanchiao, the compression ratios observed in the end-to-end example are likely to be due to two factors:
1. increasing the number of unique weights due to batch normalization and per-channel quantization after the Conv2D layer with more than 1 channel
2. zip compression not working particularly well on quantized models in general - in our internal experiments, for simple quantized (non-clustered) models, we observed nearly no compression at all!

Reply:
Thanks @akarmi for the explanation, particularly on point 2. 

Resolving this since Aron has added the one specific to clustering (1.) in the bullet point.